enhancements/installer/feature-sets.md

```diff
@@ -0,0 +1,209 @@
+4. Admin adds `featureSet: TechPreviewNoUpgrade` to install-config.yaml
+5. Installer generates `FeatureGate` manifest
+6. Cluster installs successfully with TechPreview cluster 
+
+
+### API Extensions
+
+[OpenShift API Feature Sets](https://github.com/openshift/api/blob/master/config/v1/types_feature.go)
+would become part of the install config API:
+
+```go
+
+type FeatureSet configv1.FeatureSet
+
+const (
+	// ExperimentalNoUpgrade enables features that are experimental or in active development.
+	ExperimentalNoUpgrade configv1.FeatureSet = "ExperimentalNoUpgrade"
+)
+```
+
+`ExperimentalNoUpgrade`, for Installer work in progress or
+experimental features, is added to the set of accepted feature sets.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Install Config
+
+The basic implementation through the install config is relatively straight forward. Gated features are protected
+by validation:
+
+```go
+
+	if c.FeatureSet == nil || *c.FeatureSet != types.TechPreviewNoUpgrade {
+		errMsg := "the TechPreviewNoUpgrade feature set must be enabled to use this field"
+
+    if installConfig.Foo != nil && len(installConfig.Foo.Bar) > 0 {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("foo", "bar"), errMsg))
+		}
+	}
+```
+
+Package `types` defines the install config and represents the API offered by the Installer. Most notably this
+API is consumed/used by hive. For this reason, it is important that feature sets be part of the install config.
+
+#### CLI Flag (future work)
+
+On the other hand, the install config is an asset and gated features could potentially be exercised
+before the install config is validated (or perhaps the feature is completely independent from the install config). For this reason, it would
+be a good idea to provide a flag:
+
+```shell
+$ openshift-install create cluster --feature-set TechPreviewNoUpgrade
+```
+
+This flag, if provided, would also populate the value in the install config; and conflicts between the flag and install config
+would throw an error.
+
+While there are clear use cases for a flag, there are no immediate use cases and the implementation of the Installer (with
+its separation between command-line flags and assets) does not lead to a clear-cut design. Implementing this functionality
+is certainly possible, but I propose we defer implementation for this until it is determined to be necessary.
+
+#### openshift-install explain
+
+`explain` provides a command-line reference for the install config:
+
+```shell
+$ openshift-install explain installconfig | tail -n 6
+    pullSecret <string> -required-
+      PullSecret is the secret to use when pulling images.
+
+    sshKey <string>
+      SSHKey is the public Secure Shell (SSH) key to provide access to instances.
+```
+
+`explain` should handle feature-gated fields appropriately, either by skipping them (not printing them) or clearly marking
```

Comment:
Clearly mark them as gated and fail if specified without the gate specified.

Reply:
added to latest update